File: Bucketsort/bucketsort.js
// Función de ordenamiento por inserción
function insertionSort(bukt) {
    for (let j = 1; j < bukt.length; j++) {
        const val = bukt[j];
        let k = j - 1;
        while (k >= 0 && bukt[k] > val) {
            bukt[k + 1] = bukt[k];
            k--;
        }
        bukt[k + 1] = val;
    }
    return bukt;
}

// Función principal de ordenamiento bucket
function bucketSort(inputArr) {
    const s = inputArr.length;
    if (s === 0) {
        return inputArr;
    }
    
    // Crear lista de buckets vacíos
    const bucketArr = Array.from({ length: s }, () => []);

    // Colocar cada elemento en su bucket
    for (let j of inputArr) {
        const bi = Math.floor(s * j);
        const index = bi === s ? s - 1 : bi; // Asegurar índice
        bucketArr[index].push(j);
    }

    // Ordenar cada bucket
    for (let bukt of bucketArr) {
        insertionSort(bukt);
    }

    // Concatenar los buckets en el arreglo original
    let idx = 0;
    for (let bukt of bucketArr) {
        for (let j of bukt) {
            inputArr[idx++] = j;
        }
    }
    
    return inputArr;
}

// Ejemplo de uso
const inputArr = [0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67];

console.log("Arreglo antes de ordenar:");
console.log(inputArr.join(" "));

bucketSort(inputArr);

console.log("Arreglo después de ordenar:");
console.log(inputArr.join(" "));

File: Bucketsort/Bucketsort.py
def insertion_sort(bukt):  # Función de ordenamiento por inserción
    for j in range(1, len(bukt)):  # Iterar desde el segundo elemento
        val = bukt[j]  # Elemento actual a insertar
        k = j - 1  # Índice del elemento anterior
        while k >= 0 and bukt[k] > val:  # Mover elementos mayores hacia la derecha
            bukt[k + 1] = bukt[k]  # Desplazar elemento
            k -= 1  # Mover al siguiente elemento a la izquierda
        bukt[k + 1] = val  # Insertar el elemento en su posición correcta

def bucket_sort(inputArr):  # Función principal de ordenamiento bucket
    s = len(inputArr)  # Número de buckets
    bucketArr = [[] for _ in range(s)]  # Crear lista de buckets vacios
    # Colocar cada elemento en su bucket correspondiente
    for j in inputArr:  # Iterar sobre cada elemento del arreglo
        bi = int(s * j)  # Índice del bucket
        # Asegurar que el índice del bucket no exceda el tamaño
        bucketArr[bi].append(j)  # Añadir elemento al bucket
    # Ordenar cada bucket individualmente usando ordenamiento por inserción
    for bukt in bucketArr:  # Iterar sobre cada bucket
        insertion_sort(bukt)  # Ordenar el bucket
    # Concatenar todos los buckets ordenados en el arreglo original
    idx = 0  # Índice para el arreglo original
    for bukt in bucketArr:  # Iterar sobre cada bucket
        for j in bukt:  # Iterar sobre cada elemento del bucket
            inputArr[idx] = j  # Colocar elemento en el arreglo original
            idx += 1  # Mover al siguiente índice

# Ejemplo de uso
inputArr = [0.77, 0.16, 0.38, 0.25, 0.71, 0.93, 0.22, 0.11, 0.24, 0.67]  # Arreglo de entrada
print("Arreglo antes de ordenar:")  # Imprimir arreglo antes de ordenar
print(" ".join(map(str, inputArr)))  # Imprimir elementos del arreglo
bucket_sort(inputArr)  # Llamar a la función de ordenamiento bucket
print("Arreglo después de ordenar:")  # Imprimir arreglo después de ordenar
print(" ".join(map(str, inputArr)))  # Imprimir elementos del arreglo ordenado